lib/Conversion/MooreToCore/MooreToCore.cpp

```diff
@@ -1174,6 +1174,111 @@ struct ShrOpConversion : public OpConversionPattern<ShrOp> {
+
+  LogicalResult
+  matchAndRewrite(PowUOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Type resultType = typeConverter->convertType(op.getResult().getType());
+
+    Location loc = op.getLoc();
+    auto intType = cast<IntType>(op.getRhs().getType());
+    if (auto baseOp = op.getLhs().getDefiningOp<ConstantOp>()) {
+      if (baseOp.getValue() == 2) {
+        Value constOne = rewriter.create<ConstantOp>(loc, intType, 1);
+        rewriter.replaceOpWithNewOp<ShlOp>(op, constOne, op.getRhs());
+        return success();
+      }
+    }
+
+    // transform a ** b into scf.for 0 to b step 1 { init *= a }, init = 1
+    Type integerType = rewriter.getIntegerType(intType.getWidth());
+    Value lowerBound = rewriter.create<hw::ConstantOp>(loc, integerType, 0);
+    Value upperBound =
+        rewriter.create<ConversionOp>(loc, integerType, op.getRhs());
+    Value step = rewriter.create<hw::ConstantOp>(loc, integerType, 1);
+
+    Value initVal = rewriter.create<hw::ConstantOp>(loc, resultType, 1);
+    Value lhsVal = rewriter.create<ConversionOp>(loc, resultType, op.getLhs());
+
+    auto forOp = rewriter.create<scf::ForOp>(
+        loc, lowerBound, upperBound, step, ValueRange(initVal),
+        [&](OpBuilder &builder, Location loc, Value i, ValueRange iterArgs) {
+          Value loopVar = iterArgs.front();
+          Value mul = rewriter.create<comb::MulOp>(loc, lhsVal, loopVar);
+          rewriter.create<scf::YieldOp>(loc, ValueRange(mul));
+        });
+
+    rewriter.replaceOp(op, forOp.getResult(0));
+
+    return success();
+  }
+};
+
+struct PowSOpConversion : public OpConversionPattern<PowSOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(PowSOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Type resultType = typeConverter->convertType(op.getResult().getType());
+
+    Location loc = op.getLoc();
+    auto intType = cast<IntType>(op.getRhs().getType());
+    if (auto baseOp = op.getLhs().getDefiningOp<ConstantOp>()) {
+      if (baseOp.getValue() == 2) {
+        Value constOne = rewriter.create<ConstantOp>(loc, intType, 1);
+        Value constZero = rewriter.create<ConstantOp>(loc, intType, 0);
+        Value shift = rewriter.create<ShlOp>(loc, constOne, op.getRhs());
+        Value isNegative = rewriter.create<SltOp>(loc, op.getRhs(), constZero);
+        auto condOp = rewriter.replaceOpWithNewOp<ConditionalOp>(
+            op, op.getLhs().getType(), isNegative);
+        Block *thenBlock = rewriter.createBlock(&condOp.getTrueRegion());
+        rewriter.setInsertionPointToStart(thenBlock);
+        rewriter.create<YieldOp>(loc, constZero);
+        Block *elseBlock = rewriter.createBlock(&condOp.getFalseRegion());
+        rewriter.setInsertionPointToStart(elseBlock);
+        rewriter.create<YieldOp>(loc, shift);
+        return success();
+      }
+    }
```

Comment:
This would be fantastic to have as a canonicalization in the Moore dialect! Could you move this into the canonicalizer of `moore::PowSOp`? (You might have to add a `let hasCanonicalizeMethod = 1` to the `PowSOp` definition in the `MooreOps.td` file.)

Reply:
Thanks again for your review. 

I just wanted to let you know that I followed this suggestion. 
> @maerhart : Unfortunately, canonicalization patterns and conversion patterns are not compatible, so you also need to replicate the existing canonicalization patterns functionality inside the conversion pattern.

The functions `PowSOp::canonicalize` and `PowUOp::canonicalize` already exist, and the two sections of code mentioned above were duplicated from them.